src/celeritas/io/RootStepWriter.cc

```diff
@@ -49,11 +50,27 @@ RootStepWriter::RootStepWriter(SPRootFileManager root_manager,
     , root_manager_(root_manager)
     , particles_(particle_params)
     , selection_(selection)
+    , filter_([](RootStepWriter::TStepData const&) { return true; })
 {
     CELER_EXPECT(root_manager_);
     this->make_tree();
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Construct writer with user-defined data filtering.
+ */
+RootStepWriter::RootStepWriter(SPRootFileManager root_manager,
+                               SPParticleParams particle_params,
+                               StepSelection selection,
+                               WriteFilter filter)
+    : RootStepWriter(std::move(root_manager),
+                     std::move(particle_params),
+                     std::move(selection))
+{
+    filter_ = filter;
```

Comment:
This is backward: instead of creating a default filter and overwriting it, make *this* one the primary constructor, and have the one with "default values" pass those into this constructor.